File: backend/app/schemas/graph.py
"""Pydantic schemas for graph endpoints."""
from typing import Any
from pydantic import BaseModel


class GraphNode(BaseModel):
    id: str
    type: str          # Service | Module | File | Api | Database
    label: str
    properties: dict[str, Any] = {}


class GraphEdge(BaseModel):
    id: str
    source: str
    target: str
    type: str          # CALLS | IMPORTS | DEPENDS_ON | READS | WRITES
    properties: dict[str, Any] = {}


class GraphResponse(BaseModel):
    nodes: list[GraphNode]
    edges: list[GraphEdge]
    repo_id: str
    node_count: int
    edge_count: int


class NodeDetailResponse(BaseModel):
    node: GraphNode
    neighbors: list[GraphNode]
    incoming_edges: list[GraphEdge]
    outgoing_edges: list[GraphEdge]

File: frontend/src/types/graph.ts
export type NodeType = "Service" | "Module" | "File" | "Api" | "Database";
export type EdgeType = "CALLS" | "IMPORTS" | "DEPENDS_ON" | "READS" | "WRITES" | "CONTAINS";

export interface GraphNode {
  id: string;
  type: NodeType;
  label: string;
  properties: Record<string, unknown>;
}

export interface GraphEdge {
  id: string;
  source: string;
  target: string;
  type: EdgeType;
  properties: Record<string, unknown>;
}

export interface GraphData {
  nodes: GraphNode[];
  edges: GraphEdge[];
  repo_id: string;
  node_count: number;
  edge_count: number;
}

export interface NodeDetail {
  node: GraphNode;
  neighbors: GraphNode[];
  incoming_edges: GraphEdge[];
  outgoing_edges: GraphEdge[];
}
